lib/LANraragi/Model/Tankoubon.pm

```diff
@@ -0,0 +1,270 @@
+        $redis->quit;
+        return 0;
+    }
+
+    if ( $redis->exists($tank_id) ) {
+        $redis->del($tank_id);
+        $redis->quit;
+        return 1;
+    } else {
+        $logger->warn("$tank_id doesn't exist in the database!");
+        $redis->quit;
+        return 1;
+    }
+}
+
+# update_archive_list(tankoubonid, arcid)
+#   Updates the archives list in a Tankoubon.
+#   Returns 1 on success, 0 on failure alongside an error message.
+sub update_archive_list($tank_id, $data) {
+
+    my $logger = get_logger( "Tankoubon", "lanraragi" );
+    my $redis  = LANraragi::Model::Config->get_redis;
+    my $err    = "";
+    my @tank_archives = @{$data->{"archives"}};
+
+    if ( $redis->exists($tank_id) ) {
+
+        foreach my $key (@tank_archives) {
+            unless ( $redis->exists($key) ) {
+                $err = "$key does not exist in the database.";
+                $logger->error($err);
+                $redis->quit;
+                return ( 0, $err );
+            }
+        }
+
+        my @origs = $redis->zrangebyscore($tank_id, 1, "+inf");
+        my @diff = array_difference(\@tank_archives, \@origs);
+        my @update;
+
+        # Remove the ones not in the order
+        if (@diff) {
+            $redis->zrem($tank_id, @diff);
+        }
+
+        # Prepare zadd array
+        my $len = @tank_archives;
+
+        for (my $i = 0; $i < $len; $i = $i + 1) {
+            push @update, $i+1;
+            push @update, $tank_archives[$i];
+        }
+
+        # Update
+        $redis->zadd( $tank_id, @update);
+
+        $redis->quit;
+        return ( 1, $err );
+    }
+
+    $err = "$tank_id doesn't exist in the database!";
+    $logger->warn($err);
+    $redis->quit;
+    return ( 0, $err );
+}
+
+# add_to_tankoubon(tankoubonid, arcid)
+#   Adds the given archive ID to the given Tankoubon.
+#   Returns 1 on success, 0 on failure alongside an error message.
+sub add_to_tankoubon( $tank_id, $arc_id ) {
+
+    my $logger = get_logger( "Tankoubon", "lanraragi" );
+    my $redis  = LANraragi::Model::Config->get_redis;
+    my $err    = "";
+
+    if ( $redis->exists($tank_id) ) {
+
+        unless ( $redis->exists($arc_id) ) {
+            $err = "$arc_id does not exist in the database.";
+            $logger->error($err);
+            $redis->quit;
+            return ( 0, $err );
+        }
+
+        if ($redis->zscore($tank_id, $arc_id)) {
+            $err = "$arc_id already present in category $tank_id, doing nothing.";
+            $logger->warn($err);
+            $redis->quit;
+            return ( 1, $err );
+        }
+
+        my $score = $redis->zcard($tank_id);
+
+        $redis->zadd($tank_id, $score, $arc_id);
+
+        $redis->quit;
+        return ( 1, $err );
+    }
+
+    $err = "$tank_id doesn't exist in the database!";
+    $logger->warn($err);
+    $redis->quit;
+    return ( 0, $err );
+}
+
+
+# UTILS - Probably better to move to another file
+sub array_difference($array1, $array2) {
```

Comment:
Checked it, but from what I saw, your intersect returns either the intersect or full diff, mine returns only left diff, which is what I need because I'm comparing the original list of archives with the new, and the ones in the original that are not anymore in the new needs to be processed by zrem. Moved array_difference to Utils::Generic anyways.